modules/spring-config/src/main/java/org/mule/runtime/config/internal/resolvers/DependencyGraphBeanDependencyResolver.java

```diff
@@ -0,0 +1,212 @@
+  private Map<BeanWrapper, List<BeanWrapper>> visitedComponents;
+
+
+  public DependencyGraphBeanDependencyResolver(ConfigurationDependencyResolver configurationDependencyResolver,
+                                               DeclaredDependencyResolver declaredDependencyResolver,
+                                               AutoDiscoveredDependencyResolver autoDiscoveredDependencyResolver,
+                                               AbstractSpringRegistry springRegistry) {
+    this.configurationDependencyResolver = configurationDependencyResolver;
+    this.declaredDependencyResolver = declaredDependencyResolver;
+    this.autoDiscoveredDependencyResolver = autoDiscoveredDependencyResolver;
+    this.springRegistry = springRegistry;
+    this.visitedComponentsForBuckets = new HashMap<>();
+    this.visitedComponents = new HashMap<>();
+  }
+
+  /**
+   * Default method used in other sorters to resolve bean dependencies
+   * 
+   * @param beanName the name of the bean to resolve dependencies
+   * @return
+   */
+  @Override
+  public List<Object> resolveBeanDependencies(String beanName) {
+    Object currentObject = springRegistry.get(beanName);
+    final DependencyNode currentNode = new DependencyNode(currentObject);
+
+    addDirectDependency(beanName, currentObject, currentNode, new HashSet<>());
+
+    return currentNode.getChildren()
+        .stream()
+        .map(DependencyNode::getObject)
+        .collect(toList());
+  }
+
+  @Override
+  public ConfigurationDependencyResolver getConfigurationDependencyResolver() {
+    return configurationDependencyResolver;
+  }
+
+  /**
+   * Provides only direct dependencies/required components for the object provided
+   * 
+   * @return direct children(required components) of the current object If the component is already visited for the current
+   *         bucket, we don't need to get dependencies again, return emptyList If the component is already visited for any other
+   *         buckets, but not for this bucket, return the dependencies already saved If it was never visited before, get direct
+   *         dependencies for the object
+   */
+  public List<BeanWrapper> getDirectBeanDependencies(BeanWrapper currentObject, int bucketIndex) {
+    // If the component was already processed for the current bucket(graph), it won't process it
+    visitedComponentsForBuckets.putIfAbsent(bucketIndex, new HashSet<>());
+    if (visitedComponentsForBuckets.get(bucketIndex).stream().map(x -> x.getWrappedObject())
+        .anyMatch(x -> x.equals(currentObject.getWrappedObject()))) {
+      return emptyList();
+    }
+    visitedComponentsForBuckets.get(bucketIndex).add(currentObject);
+
+    // If the component was already processed for any other buckets,
+    // it returns the direct dependencies that were saved before to build a dependency graph for the current bucket
+    if (visitedComponents.containsKey(currentObject)) {
+      return visitedComponents.get(currentObject);
+    }
```

Comment:
Buckets define an order that the graph does not. Some objects of certain types should be initialized before other objects of other types. We have no strategy to generate an edge for these dependencies so we use a bucket sort. Now, when we get an object as an input, we find the bucket it belongs to. Now we need to know which object this one depends on regardless of the bucket. Maybe this second object is actually a dependency for many other objects as well. Some of those might not end up in this bucket.
Each bucket has its own objects to take care of and uses a graph to do so. If an object "A" shows up many times as a dependency, when we get "A"'s dependencies we would get the same list over and over again so we try to do it only once and save the result to reuse it.

Reply:
The original version creates buckets for classes that should be initialized and checked with for loop if each object from the lookup is the instance of the bucket. If object A is an instance of bucket 3 and 11, it will be in bucket 3 in the original implementation. After ordering them, all the required prerequisites for each object will be initialized repeatedly regardless of their existence. This is where we get 80K repeated objects during initialization phase.
 
At first, we made a single graph for each phase, the order of allowed classes still matters so we reintroduce the buckets. Also, there were some objects like MuleContext that could be the prerequisite of many objects across the buckets, even in the same bucket. Caching the result and checking the existence in the bucket helps reduce repetition.

Reply:
Ok, I think what confused me was not knowing whether we would be returning all dependencies regardless of whether they belong to the same bucket or even the same phase. Based on what you are saying it seems we are returning all of them.
So if we are on the initialization phase, and A is initializable and depends on B, but B is not initializable, we would be returning B as dependency anyway (even though we wouldn't need to initialize it). Is that right?
For this reason, the dependencies would work for all buckets (and maybe all phases too?).

Reply:
Yes, if A(initializable)->B(not initializable), then we will return that information as well because of the edge case A->B->C(initializable). In the case of A->B->C, we need to have the information that A indirectly relies C during the phase.